src/opensynth/models/faraday/model.py

```diff
@@ -372,26 +477,29 @@ def sample_gmm(
               Decoder output (KWH), month label, dow label
         """
         gmm_samples = self.gmm.sample(n_samples)[0]
+
+        # Parse labels and profiles
         gmm_kwh = gmm_samples[:, : self.vae_module.latent_dim]
-        gmm_mth = np.round(gmm_samples[:, -2], decimals=0).astype(int)
-        gmm_dow = np.round(gmm_samples[:, -1], decimals=0).astype(int)
+        gmm_labels: dict[str, torch.tensor] = {}
+        # Abstract the labels and round numerical values to integers
+        for i, feature in enumerate(self.vae_module.feature_list):
+            index = self.get_index(self.vae_module.feature_list, i)
+            gmm_labels[feature] = np.round(
+                gmm_samples[:, index], decimals=0
+            ).astype(int)
```

Comment:
I am not sure I understand why this code has to be in a for-loop? 
From my understanding, `get_index` just converts indices [0,1,2,3] -> [-4,-3,-2,-1] s.t. it can access the features in `gmm_samples`, so I propose `get_index` returns an array (or tensor), r.t. an integer, given by `-(len(self.vae_module.feature_list) - range(0, len(self.vae_module.feature_list), 1)[::-1]` . This avoids the need for a for-loop.

Reply:
Also the rounding step can be taken out the for-loop (assuming that you can do `np.round(gmm_samples[:, list_indicies], decimals=0).astype(int)`)

Reply:
It's not just rounding, but also appending to the `gmm_labels` dictionary in an ordered manner, so that it can be decoded by the VAE.  The order of the dictionary needs to be preserved, and that check is handled in the `train_gmm` step (to check that the data module used to train the GMM module loads in features in the same order as seen by VAE). Otherwise you're sampling/ conditioning on the wrong labels. 

So TL;DR the for loop can't be avoided. It needs to iterate through the label columns so that it can be stored in a python dictionary in the right order. And if the for-loop can't be avoided, I don't quite see the value of taking the rounding step out of the for-loop.

The alternative implementation of taking the the rounding out of the for-loop without the use of `get_index` as you suggested would be:
```python
  gmm_kwh: torch.Tensor = gmm_samples[:, : self.vae_module.latent_dim]

  gmm_label_int: npt.NDArray[np.int64] = np.round(
      gmm_samples[:, self.vae_module.latent_dim :], decimals=0
  )
  gmm_labels: dict[str, torch.Tensor] = {}

  for i, feature in enumerate(self.vae_module.feature_list):
      gmm_labels[feature] = gmm_label_int[:, i]
```

But I don't think this implementation is solving anything and in fact it makes the code less explicit as the index from `for i, feature in enumerate(self.vae_module.feature_list):` is no longer coupled explicitly with the index of `gmm_label_int` array. To unit test the the logic, we then have to wrap this in a much larger function and test more things.

It's more unit testable I think to abstract the logic in the `get_index` function so we can just test it explicitly. There's a bug with the pytest however - fixing in latest commit.

Reply:
Okay, I can see why the loop is simpler/clearer/easier for testing. Thanks for the response. Ideally, should this for loop also be in a function so it can be unit tested to ensure `gmm_labels` are as expected?

Reply:
We could put the for loop too in a function to test it, but depends on how we are writing the test for it!